### PR TITLE
cache: Adds support for caching HTTP trailers.

### LINKED ISSUE #9975: CacheFilter: Support Trailers
CacheFilter should support caching HTTP trailers.

This doesn't block widespread deployment of CacheFilter, and won't be a priority unless someone pushes for it.

### PR DESCRIPTION
**Pull request by ekiziv**
Signed-off-by: Eleonora Kiziv <ekiziv@google.com>

Commit Message: cache: Adds support for caching HTTP trailers. 
Risk Level: Low
Testing: Unit, integration test
Fixes: #9975

### CODE REVIEW

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (jmarantz) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:26]: from a stylistic perspective I don't think Envoy uses this `/*has_trailers=*/` convention, though I'm fine with it myself.

**Review by jmarantz — COMMENTED** [review]
On the whole this looks good but can you get a first pass from someone working in the caching code currently?

Thanks!

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (capoferro) [inline test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:280]: I know the rest of the tests have request path variables like this, and so you are staying consistent with the file, but we should provide a real path in the `RequestPath#` vars, rather than just an arbitrary string, ie. with a leading slash: `"/name"`. This has no functional impact right now, but it will make a difference as we add more features in the near future.

The rest of the file is wrong, and is something we should fix at some point, while simultaneously not adding more with the same mistake.

**Review by capoferro — CHANGES_REQUESTED** [review]
Looks great! One minor comment wrt path variables in the test, but otherwise 👍

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (jmarantz) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:135]: Nit: if you rewrite this as an if-statement, it will be the same number of lines, but might be easier to read.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (jmarantz) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:174]: ditto

File: test/extensions/filters/http/cache/cache_filter_integration_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/cache/cache_filter_integration_test.cc:325]: nit: compare the age to "" and then if it fails the test log will tell you what the age actually was, rather just the size of the string.

File: test/extensions/filters/http/cache/cache_filter_integration_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/cache/cache_filter_integration_test.cc:334]: is this intended to remain?

File: test/extensions/filters/http/cache/cache_filter_integration_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/cache/cache_filter_integration_test.cc:344]: ditto

File: test/extensions/filters/http/cache/cache_filter_integration_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/cache/cache_filter_integration_test.cc:366]: nit: this pattern was pre-existing in this file, but you can make an empty optref with:
```
OptRef<const std::string>()
```
here and below.

You could also just declare
```
  OptRef<const std::string> empty_body_;
  OptRef<const Http::TestResponseTrailerMapImp> empty_trailers_;
```
in the test class and pass that in, to avoid the `/*response_body=*/` syntax, just to make the call-sites more concise.

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (jmarantz) [reply test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:280]: also I don't think we use the RequestPath1 spelling for local consts. The Envoy style guide is not specific but it points to the Google style guide which says:

> and whose value is fixed for the duration of the program

So I think just name this `request_path1`.

**Review by jmarantz — COMMENTED** [review]
mostly just nits: i'm counting on @capoferro to provide the detaile semantic review.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (ekiziv) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:26]: Removed.

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (ekiziv) [reply test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc]: Hi! Thank you for your comments! Would be happy to update the path, but I am not sure to what value. So, we could leave it like this for now or if you give me a little more context on what it should include(there does not seem to be consensus in references), I could update it.

File: test/extensions/filters/http/cache/cache_filter_integration_test.cc
Comment (ekiziv) [reply test/extensions/filters/http/cache/cache_filter_integration_test.cc:324]: Please, see this discussion on why we decided to check the size here: https://github.com/envoyproxy/envoy/pull/15910#discussion_r612404635 and let me know if you want me to update it further.

File: test/extensions/filters/http/cache/cache_filter_integration_test.cc
Comment (ekiziv) [reply test/extensions/filters/http/cache/cache_filter_integration_test.cc]: Uuu, this is a great suggestion! Thank you!

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (jmarantz) [reply test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc]: Suggest:
```
const std::string request_path1 = "/path";
```

File: test/extensions/filters/http/cache/cache_filter_integration_test.cc
Comment (jmarantz) [reply test/extensions/filters/http/cache/cache_filter_integration_test.cc:324]: Oh sorry ... yeah what you have is correct. Optional suggestion to make it slightly more obvious what's going on:
```
  Http::GetResult age_header_values = response_decoder->headers().get(Http::CustomHeaders::get().Age);
  EXPECT_EQ(0, age_header_values.size());
```
Up to you whether you think the readability justifies the extra line.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (jmarantz) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:135]: you don't need to init this; unique_ptr has a constructor which has the same effect.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (jmarantz) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:171]: you don't need to init this; unique_ptr has a constructor which has the same effect.

**Comment by ekiziv** [comment]
I am reopening this pull request, as I have made the necessary modifications.

File: test/extensions/filters/http/cache/http_cache_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/cache/http_cache_test.cc:390]: nit: assert lookup_response.headers_ != nullptr

File: test/extensions/filters/http/cache/http_cache_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/cache/http_cache_test.cc:393]: i'm trying to follow this test flow and it's not obvious to me where the age gets populated. Can you add some comments to the code explaining this?

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:108]: Can you rename this expectLookupSuccessWIthBodyAndTrailers? I also don't think you necessarily want the above if-statement. If the caller doesn't supply trailers we should go ahead and check they are empty, no?

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:298]: See above comment about leaving expectLookupSuccessWithBody as it was, and testing the trailers separately. In this case you'd need a temp for the lookup result so there'd be one extra line:
```
LookupContextPtr lookup_context = lookup(request_path);
expectLookupSuccessWithBody(lookup_context.get(), new_body);
EXPECT_EQ(response_trailers, getTrailers(*lookup_context));
```

Same with the call-site on line 289 though in that case there'd be 2 extra lines.

**Review by jmarantz — COMMENTED** [review]
Can someone on the caching team also look at this for better insight into caching semantics?

**Comment by yanavlasov** [comment]
@toddmgreer @penguingao @mpwarres @capoferro for final review.

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (jmarantz) [reply test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:298]: Sorry, this comment is obsolete; I started to make this suggestion in another comment but changed my mind, and then forgot to delete this comment.

File: test/extensions/filters/http/cache/http_cache_test.cc
Comment (ekiziv) [reply test/extensions/filters/http/cache/http_cache_test.cc]: Done.

File: test/extensions/filters/http/cache/http_cache_test.cc
Comment (ekiziv) [reply test/extensions/filters/http/cache/http_cache_test.cc:394]: I attempted to explain where it is set. Let me know if there is something specific I should mention.

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (ekiziv) [reply test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc]: Done.

**Comment by capoferro** [comment]
LGTM. Semantically this is correct. Tyvm @jmarantz for the detailed review.

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (capoferro) [inline test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:278]: This string should be something like `"/name"` so it provides a valid path (including the leading slash).

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (capoferro) [inline test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:140]: Same as below: This string should be something like "/name" so it provides a valid path (including the leading slash).

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (capoferro) [inline test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:128]: Same as below: This string should be something like "/name" so it provides a valid path (including the leading slash).

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (jmarantz) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:38]: it is unexpected that cb() would simply not be called if trailers is null. In contrast, for the getHeaders case, cb is called regardless, though sometimes wiht an empty LookupResult (that also seems potentially confusing).

Maybe it'd be more obvious what the semantics are to call with a std::optional<LookupResult> in both cases?

File: test/extensions/filters/http/cache/http_cache_test.cc
Comment (jmarantz) [reply test/extensions/filters/http/cache/http_cache_test.cc:394]: Got it, thanks.

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:110]: drop these debug statements

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:252]: s/RequestPath/request_path/. I realize the style guide says to use RequestPath for constants, but I'm pretty sure the intent is to use that for globals, not for local vars. Elsewhere in Envoy code we are not using that spelling for local const variables.

**Review by jmarantz — COMMENTED** [review]
sorry for the delay; I had a bit of vacation interrupt the flow of this review.

Making great progress; still have a few questions/nits.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (ekiziv) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc]: I implemented the same check as in getHeaders for consistency, but I could change the semantics in the subsequent CL if you think it would still be worth it.

**Comment by ekiziv** [comment]
Thanks for your suggestions and patience!

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (jmarantz) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:37]: Do we have a testcase that hits the case where trailers is not present?

This was pre-existing but it'd be good to make sure we also have a case for where headers are not present and so the getHeaders callback gets called with am empty LookupResult.

Also, WDYT of using absl::optional rather than empty-constructed objects for both those cases?

**Comment by jmarantz** [comment]
needs main merge

/wait

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (toddmgreer) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:37]: jmarantz, I think you're suggesting changing LookupTrailersCallback's parameter from ResponseTrailerMapPtr&& to optional<ResponseTrailerMap>&&. That would cause an unnecessary copy, because these trailers get passed to StreamDecoderFilterCallbacks::encodeTrailers, which requires ResponseTrailerMapPtr&&.

LookupTrailerCallback should take the same parameter type as StreamDecoderFilterCallbacks::encodeTrailers does.

Comment (toddmgreer) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:37]: The declaration comment of HttpCache::getTrailers should say something like "The cache must call cb with the cached trailers." (Sorry for the omission; feel free to add that if so inclined.) I suggest something like:

void getTrailers(LookupTrailersCallback&& cb) override {
  ASSERT(trailers_);
  cb(Http::createHeaderMap<Http::ResponseTrailerMapImpl>(*trailers_));
}

CacheFilter promises to only call getTrailers if LookupResult::has_trailers_ is true. If getTrailers gets called while trailers_ is null, it's a bug.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (toddmgreer) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:36]: There's no need to create a new ResponseTrailerMap. `cb(std::move(trailers_))` is simpler and more efficient.

**Review by toddmgreer — CHANGES_REQUESTED** [review]

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (ekiziv) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc]: Thanks for your suggestion, Todd!

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (jmarantz) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc]: Thanks Todd -- keeping the type as is SGTM.

There's still my pending question "Do we have a testcase that hits the case where trailers is not present?"

Comment (ekiziv) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc]: Sorry, yes, I have added the requested test case in https://github.com/envoyproxy/envoy/pull/16854/commits/df64f002894465e01a1fb695c14de82e106710e6.

**Comment by ekiziv** [comment]
You might have missed the comment above, but I have added the requested test case in https://github.com/envoyproxy/envoy/commit/df64f002894465e01a1fb695c14de82e106710e6. Let me know if there is something else missing.

**Review by jmarantz — APPROVED** [review]
Thanks, looks great!

/assign-from @envoyproxy/senior-maintainers

**Comment by alyssawilk** [comment]
Sorry about this staling out waiting on review.  as this is extension-only code @jmarantz can simply merge it himself if you resolve the conflicts
/wait

**Comment by capoferro** [comment]
The CI failures may not be related to these changes... the logs say the tests pass, but that a file was missing after the test run: [297456](https://dev.azure.com/cncf/envoy/_build/results?buildId=98274&view=results)

**Comment by capoferro** [comment]
I take it back, it is a real failure. Looks like a variable name hiccup post-merge.

**Comment by capoferro** [comment]
Looks like a merge hiccup. I'll reach out and help sort this out.

/wait

**Comment by ekiziv** [comment]
@dio, @htuch, @lizan, @snowp, @zuercher, you have been automatically added to this PR due to a merge mishap. Sorry about that and you could remove yourself from reviewers. Thank you!

**Review by jmarantz — COMMENTED** [review]
This looks fine to me now as maintainer, but can we get LGTM from @capoferro ?

**Review by capoferro — APPROVED** [review]
LGTM!

**Review by jmarantz — APPROVED** [review]

**Pull request merged**
